cpp/src/sort/segmented_sort.cu

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/table/table_device_view.cuh>
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+//#include <cub/device/device_segmented_radix_sort.cuh>
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/logical.h>
+
+namespace cudf {
+namespace detail {
+
+// returns segment indices for each element for all segments.
+rmm::device_uvector<size_type> get_segment_indices(size_type num_rows,
+                                                   column_view const& offsets,
+                                                   rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<size_type> segment_ids(num_rows, stream);
+
+  auto offset_begin = offsets.begin<size_type>();  // assumes already offset column contains offset.
+  auto offsets_minus_one = thrust::make_transform_iterator(
+    offset_begin, [offset_begin] __device__(auto i) { return i - offset_begin[0] - 1; });
+  auto counting_iter = thrust::make_counting_iterator<size_type>(0);
+  thrust::lower_bound(rmm::exec_policy(stream),
+                      offsets_minus_one + 1,
+                      offsets_minus_one + offsets.size(),
+                      counting_iter,
+                      counting_iter + segment_ids.size(),
+                      segment_ids.begin());
+  return std::move(segment_ids);
+}
+
+// same as count_elements(list). TODO: DRY.
```

Comment:
Intentionally leaving this in?

Reply:
@davidwendt can this code be moved to a .cuh file and reuse it in `count_elements.cu`?

Reply:
Certainly. My suggestion is to move into the existing `list_device_view.cuh`
Seems like we should be able to factor the logic into there sometime in the future.

Reply:
moved.